plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,698 @@
+    {
+        if (uri.getHost() != null) {
+            return uri.getHost();
+        }
+
+        if (uri.getUserInfo() == null) {
+            return uri.getAuthority();
+        }
+
+        throw new IllegalArgumentException("Unable to determine S3 bucket from URI.");
+    }
+
+    private AmazonS3 createAmazonS3Client(ClientConfiguration clientConfig, AWSCredentialsProvider credentialsProvider)
+    {
+        AmazonS3Builder<AmazonS3ClientBuilder, AmazonS3> clientBuilder = AmazonS3Client.builder()
+                .withCredentials(credentialsProvider)
+                .withClientConfiguration(clientConfig);
+
+        boolean regionOrEndpointSet = false;
+
+        if (pinS3ClientToCurrentRegion) {
+            clientBuilder.setRegion(Regions.getCurrentRegion().getName());
+            regionOrEndpointSet = true;
+        }
+
+        if (endPoint != null) {
+            clientBuilder.setEndpointConfiguration(new EndpointConfiguration(endPoint, null));
+            regionOrEndpointSet = true;
+        }
+
+        if (!regionOrEndpointSet) {
+            clientBuilder.withRegion(US_EAST_1);
+            clientBuilder.setForceGlobalBucketAccessEnabled(true);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private static AWSCredentialsProvider createAwsCredentialsProvider(ExchangeS3Config config)
+    {
+        if (config.getS3AwsAccessKey() != null && config.getS3AwsSecretKey() != null) {
+            return new AWSStaticCredentialsProvider(
+                    new BasicAWSCredentials(config.getS3AwsAccessKey(), config.getS3AwsSecretKey()));
+        }
+        if (config.getS3IamRole() != null) {
+            return new STSAssumeRoleSessionCredentialsProvider
+                    .Builder(config.getS3IamRole(), "trino-session")
+                    .withExternalId(config.getS3ExternalId())
+                    .build();
+        }
+        return DefaultAWSCredentialsProviderChain.getInstance();
+    }
+
+    private static boolean isDirectory(URI uri)
+    {
+        return uri.toString().endsWith(PATH_SEPARATOR);
+    }
+
+    private static class S3ExchangeSliceOutput
+            extends SliceOutput
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(S3ExchangeSliceOutput.class).instanceSize();
+
+        private final AmazonS3 s3;
+        private final String bucketName;
+        private final String key;
+        private final Supplier<String> uploadIdFactory;
+        private final ExecutorService uploadExecutor;
+        private final Optional<SSECustomerKey> sseCustomerKey;
+
+        private int currentPartNumber;
+        private byte[] buffer;
+        private int bufferSize;
+
+        private boolean failed;
+        private Optional<String> uploadId = Optional.empty();
+        private Future<UploadPartResult> inProgressUploadFuture;
+        private final List<UploadPartResult> parts = new ArrayList<>();
+
+        // TODO: implement async upload (aka. double buffering)
```

Comment:
Not sure if this is the right level to implement async upload (double buffering). I wonder if it would be better to implement it at the `ExchangeSink` level. That way the mechanism could be reused for all future supported file systems at no additional cost.